src/box/txn.c

```diff
@@ -553,18 +555,33 @@ txn_begin_in_engine(struct engine *engine, struct txn *txn)
 {
 	if (engine->flags & ENGINE_BYPASS_TX)
 		return 0;
-	if (txn->engine == NULL) {
-		txn->engine = engine;
-		return engine_begin(engine, txn);
-	} else if (txn->engine != engine) {
-		/**
-		 * Only one engine can be used in
-		 * a multi-statement transaction currently.
-		 */
-		diag_set(ClientError, ER_CROSS_ENGINE_TRANSACTION);
-		return -1;
+	if (txn->engines[engine->id] != NULL) {
+		assert(txn->engines[engine->id] == engine);
+		return 0;
+	}
+	txn->engines_count++;
+	txn->engines[engine->id] = engine;
+
+	if (txn->engines_count > 1) {
+		for (size_t i = 0; i < MAX_TX_ENGINE_COUNT; i++) {
+			struct engine *engine = txn->engines[i];
+			if (engine != NULL &&
+			    !(engine->flags &
+			      ENGINE_SUPPORTS_CROSS_ENGINE_TX)) {
+				diag_set(ClientError,
+					 ER_CROSS_ENGINE_TRANSACTION,
+					 engine->name);
+				goto error;
+			}
+		}
 	}
```

Comment:
I don't like this check. Let's rework.
1. Add two boolean flags to `txn_flag`: `TXN_IS_STARTED` and `TXN_SUPPORTS_MULTI_ENGINE` (the names might need revising). `TXN_IS_STARTED` is initially false; `TXN_SUPPORTS_MULTI_ENGINE` is initially true.
2. If we add a new engine and `TXN_IS_STARTED` is false, set it to true, and clear `TXN_SUPPORTS_MULTI_ENGINE` if the engine doesn't support multi-engine transactions.
3. If we add a new engine that doesn't support multi-engine transactions and `TXN_IS_STARTED` is set, abort.
4. If we add a new engine and `TXN_SUPPORTS_MULTI_ENGINE` is unset, abort.

The we wouldn't need to iterate over all engines here.

Reply:
OK, reworked this way.